README.md
[![Latest Stable Version]](https://packagist.org/packages/wikimedia/wait-condition-loop) [![License]](https://packagist.org/packages/wikimedia/wait-condition-loop)

Wait Condition Loop for PHP
===========================

This class is used for waiting on a condition to be reached, with the ability
to specify a timeout. The condition is considered reached when the condition callback
returns CONDITION_REACHED or true. CONDITION_ABORTED can also be used to stop the loop.

Additionally, "work" callbacks can be injected to prepare useful work instead of simply
having the current thread sleep or block on I/O. The loop will run one of these callbacks
on each iteration of checking the condition callback, as long as there are any left to run.

The loop class will automatically either retry the condition or usleep() before retrying it,
depending on CPU usage. Low CPU usage and significant real-time passage is used to detect
whether the condition callback appears to use blocking I/O. Use of usleep() will not occur
until all of the "work" callbacks have run. This means that the condition callback can
either be an "instant" CPU-bound check or a blocking I/O call with a small timeout. Both
cases should automatically work without CPU intensive spin loops.

Additional documentation about the library can be found on
[MediaWiki.org](https://www.mediawiki.org/wiki/WaitConditionLoop).


Usage
-----

    // Pre-compute some value that will be needed later
    $result = null;
    $workCallback = function () use ( &$result ) {
        $result = ( $result !== null ) ? $result : $this->doWork();

        return $result
    }

    $loop = new WaitConditionLoop(
        function () use ( ... ) {
            if ( ... ) {
                // Condition reached; stop loop
                return WaitConditionLoop::CONDITION_REACHED;
            }
            // Condition not reached; keep checking
            return WaitConditionLoop::CONDITION_CONTINUE;
        },
        3.0, // timeout in seconds
        [ $workCallback ]
    );
    $status = $loop->invoke(); // CONDITION_* constant

    // Call $workCallback as needed later

Running tests
-------------

    composer install --prefer-dist
    composer test


---
[Latest Stable Version]: https://poser.pugx.org/wikimedia/wait-condition-loop/v/stable.svg
[License]: https://poser.pugx.org/wikimedia/wait-condition-loop/license.svg


HISTORY.md
# Changelog

## 2.0.1

* Remove remnant HHVM check. (Reedy)

## 2.0.0

This release requires PHP 7.2+.

* Remove support for PHP 5.5/5.6/7.0/7.1 and HHVM. (James D. Forrester)

## 1.0.0

* Initial release. (Aaron Schulz)


tests/WaitConditionLoopTest.php
<?php
/**
 * Copyright (C) 2016 Aaron Schulz <aschulz@wikimedia.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @file
 * @author Aaron Schulz <aschulz@wikimedia.org>
 */

namespace Wikimedia\WaitConditionLoop\Test;

use Wikimedia\WaitConditionLoop;

/**
 * @covers \Wikimedia\WaitConditionLoop
 */
class WaitConditionLoopTest extends \PHPUnit\Framework\TestCase {
	public function testCallbackReached() {
		$wallClock = microtime( true );

		$count = 0;
		$status = new \stdClass();
		$loop = new WaitConditionLoopFakeTime(
			static function () use ( &$count, $status ) {
				++$count;
				$status->value = 'cookie';

				return WaitConditionLoop::CONDITION_REACHED;
			},
			10.0,
			$this->newBusyWork( $x, $y, $z )
		);
		$this->assertEquals( $loop::CONDITION_REACHED, $loop->invoke() );
		$this->assertSame( 1, $count );
		$this->assertEquals( 'cookie', $status->value );
		$this->assertEquals( [ 0, 0, 0 ], [ $x, $y, $z ], "No busy work done" );

		$count = 0;
		$loop = new WaitConditionLoopFakeTime(
			static function () use ( &$count, &$wallClock ) {
				$wallClock += 1;
				++$count;

				return $count >= 2 ? WaitConditionLoop::CONDITION_REACHED : false;
			},
			7.0,
			$this->newBusyWork( $x, $y, $z, $wallClock )
		);
		$this->assertEquals( $loop::CONDITION_REACHED, $loop->invoke(),
			"Busy work did not cause timeout" );
		$this->assertEquals( [ 1, 0, 0 ], [ $x, $y, $z ] );

		$count = 0;
		$loop = new WaitConditionLoopFakeTime(
			static function () use ( &$count, &$wallClock ) {
				$wallClock += 0.1;
				++$count;

				return $count > 80 ? true : false;
			},
			50.0,
			$this->newBusyWork( $x, $y, $z, $wallClock, $dontCallMe, $badCalls )
		);
		$this->assertSame( 0, $badCalls, "Callback exception not yet called" );
		$this->assertEquals( $loop::CONDITION_REACHED, $loop->invoke() );
		$this->assertEquals( [ 1, 1, 1 ], [ $x, $y, $z ], "Busy work done" );
		$this->assertSame( 1, $badCalls, "Bad callback ran and was exception caught" );

		try {
			$e = null;
			$dontCallMe();
		} catch ( \Exception $e ) {
		}

		$this->assertInstanceOf( 'RunTimeException', $e );
		$this->assertSame( 1, $badCalls, "Callback exception cached" );
	}

	public function testCallbackTimeout() {
		$count = 0;
		$wallClock = microtime( true );
		$loop = new WaitConditionLoopFakeTime(
			static function () use ( &$count, &$wallClock ) {
				$wallClock += 3;
				++$count;

				return $count > 300 ? true : false;
			},
			50.0,
			$this->newBusyWork( $x, $y, $z, $wallClock )
		);
		$loop->setWallClock( $wallClock );
		$this->assertEquals( $loop::CONDITION_TIMED_OUT, $loop->invoke() );
		$this->assertEquals( [ 1, 1, 1 ], [ $x, $y, $z ], "Busy work done" );

		$loop = new WaitConditionLoopFakeTime(
			static function () use ( &$count, &$wallClock ) {
				$wallClock += 3;
				++$count;

				return true;
			},
			0.0,
			$this->newBusyWork( $x, $y, $z, $wallClock )
		);
		$this->assertEquals( $loop::CONDITION_REACHED, $loop->invoke() );

		$count = 0;
		$loop = new WaitConditionLoopFakeTime(
			static function () use ( &$count, &$wallClock ) {
				$wallClock += 3;
				++$count;

				return $count > 10 ? true : false;
			},
			0,
			$this->newBusyWork( $x, $y, $z, $wallClock )
		);
		$this->assertEquals( $loop::CONDITION_FAILED, $loop->invoke() );
	}

	public function testCallbackAborted() {
		$x = 0;
		$wallClock = microtime( true );
		$loop = new WaitConditionLoopFakeTime(
			static function () use ( &$x, &$wallClock ) {
				$wallClock += 2;
				++$x;

				return $x > 2 ? WaitConditionLoop::CONDITION_ABORTED : false;
			},
			10.0,
			$this->newBusyWork( $x, $y, $z, $wallClock )
		);
		$loop->setWallClock( $wallClock );
		$this->assertEquals( $loop::CONDITION_ABORTED, $loop->invoke() );
	}

	public function testLastWaitTime() {
		$list = [];
		$wallClock = microtime( true );
		$count = 0;
		$loop = new WaitConditionLoopFakeTime(
			static function () use ( &$count, &$wallClock ) {
				$wallClock += 1.0;
				$count++;
				return ( $count > 2 );
			},
			10.0,
			$list
		);
		$loop->setWallClock( $wallClock );
		$this->assertEquals( $loop::CONDITION_REACHED, $loop->invoke() );
		$this->assertEquals( 2.0, $loop->getLastWaitTime() );
	}

	public function testAbortedWithNoTimeout() {
		$loop = new WaitConditionLoop(
			static function () {
				return WaitConditionLoop::CONDITION_ABORTED;
			},
			0
		);

		$this->assertEquals( $loop::CONDITION_FAILED, $loop->invoke() );
	}

	private function newBusyWork(
		&$x, &$y, &$z, &$wallClock = 1, &$dontCallMe = null, &$badCalls = 0
	) {
		$x = $y = $z = 0;
		$badCalls = 0;

		$list = [];
		$list[] = static function () use ( &$x, &$wallClock ) {
			$wallClock += 1;

			return ++$x;
		};
		$dontCallMe = static function () use ( &$badCalls ) {
			++$badCalls;
			throw new \RuntimeException( "TrollyMcTrollFace" );
		};
		$list[] =& $dontCallMe;
		$list[] = static function () use ( &$y, &$wallClock ) {
			$wallClock += 15;

			return ++$y;
		};
		$list[] = static function () use ( &$z, &$wallClock ) {
			$wallClock += 0.1;

			return ++$z;
		};

		return $list;
	}
}


tests/WaitConditionLoopFakeTime.php
<?php
/**
 * Copyright (C) 2016 Aaron Schulz <aschulz@wikimedia.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @file
 * @author Aaron Schulz <aschulz@wikimedia.org>
 */
namespace Wikimedia\WaitConditionLoop\Test;

use Wikimedia\WaitConditionLoop;

class WaitConditionLoopFakeTime extends WaitConditionLoop {
	/**
	 * @var int
	 */
	protected $wallClock = 1;

	/**
	 * @inheritDoc
	 */
	public function __construct( callable $condition, $timeout, array $busyCallbacks ) {
		parent::__construct( $condition, $timeout, $busyCallbacks );
	}

	/**
	 * @inheritDoc
	 */
	protected function usleep( $microseconds ) {
		$this->wallClock += $microseconds / 1e6;
	}

	/**
	 * @inheritDoc
	 */
	protected function getCpuTime() {
		return 0.0;
	}

	/**
	 * @inheritDoc
	 */
	protected function getWallTime() {
		return $this->wallClock;
	}

	/**
	 * @inheritDoc
	 */
	public function setWallClock( &$timestamp ) {
		$this->wallClock =& $timestamp;
	}
}


src/WaitConditionLoop.php
<?php
/**
 * Wait loop that reaches a condition or times out.
 *
 * Copyright (C) 2016 Aaron Schulz <aschulz@wikimedia.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @file
 * @author Aaron Schulz
 */

namespace Wikimedia;

/**
 * Wait loop that reaches a condition or times out
 */
class WaitConditionLoop {
	/** @var callable */
	private $condition;
	/** @var callable[] */
	private $busyCallbacks = [];
	/** @var float Seconds */
	private $timeout;
	/** @var float Seconds */
	private $lastWaitTime;
	/** @var int|null */
	private $rusageMode;

	public const CONDITION_REACHED = 1;
	// evaluates as falsey
	public const CONDITION_CONTINUE = 0;
	public const CONDITION_FAILED = -1;
	public const CONDITION_TIMED_OUT = -2;
	public const CONDITION_ABORTED = -3;

	/**
	 * @param callable $condition Callback that returns a WaitConditionLoop::CONDITION_ constant
	 * @param float $timeout Timeout in seconds
	 * @param array &$busyCallbacks List of callbacks to do useful work (by reference)
	 */
	public function __construct( callable $condition, $timeout = 5.0, &$busyCallbacks = [] ) {
		$this->condition = $condition;
		$this->timeout = $timeout;
		$this->busyCallbacks =& $busyCallbacks;

		// @codeCoverageIgnoreStart
		if ( function_exists( 'getrusage' ) ) {
			// RUSAGE_SELF
			$this->rusageMode = 0;
		}
		// @codeCoverageIgnoreEnd
	}

	/**
	 * Invoke the loop and continue until either:
	 *   - a) The condition callback returns neither CONDITION_CONTINUE nor false
	 *   - b) The timeout is reached
	 * Thus a condition callback can return true (stop) or false (continue) for convenience.
	 * In such cases, the halting result of "true" will be converted to CONDITION_REACHED.
	 *
	 * If $timeout is 0, then only the condition callback will be called (no busy callbacks),
	 * and this will immediately return CONDITION_FAILED if the condition was not met.
	 *
	 * Exceptions in callbacks will be caught and the callback will be swapped with
	 * one that simply rethrows that exception back to the caller when invoked.
	 *
	 * @return int WaitConditionLoop::CONDITION_* constant
	 * @throws \Exception Any error from the condition callback
	 */
	public function invoke() {
		// seconds
		$elapsed = 0.0;
		// microseconds to sleep each time
		$sleepUs = 0;
		$lastCheck = false;
		$finalResult = self::CONDITION_TIMED_OUT;
		do {
			$checkStartTime = $this->getWallTime();
			// Check if the condition is met yet
			$realStart = $this->getWallTime();
			$cpuStart = $this->getCpuTime();
			$checkResult = call_user_func( $this->condition );
			$cpu = $this->getCpuTime() - $cpuStart;
			$real = $this->getWallTime() - $realStart;
			// Exit if the condition is reached, an error occurs, or this is non-blocking
			if ( $this->timeout <= 0 ) {
				$finalResult = $checkResult ? self::CONDITION_REACHED : self::CONDITION_FAILED;
				break;
			} elseif ( (int)$checkResult !== self::CONDITION_CONTINUE ) {
				if ( is_int( $checkResult ) ) {
					$finalResult = $checkResult;
				} else {
					$finalResult = self::CONDITION_REACHED;
				}
				break;
			} elseif ( $lastCheck ) {
				// timeout reached
				break;
			}
			// Detect if condition callback seems to block or if justs burns CPU
			$conditionUsesInterrupts = ( $real > 0.100 && $cpu <= $real * 0.03 );
			if ( !$this->popAndRunBusyCallback() && !$conditionUsesInterrupts ) {
				// 10 queries = 10(10+100)/2 ms = 550ms, 14 queries = 1050ms
				// stop incrementing at ~1s
				$sleepUs = min( $sleepUs + 10 * 1e3, 1e6 );
				$this->usleep( $sleepUs );
			}
			$checkEndTime = $this->getWallTime();
			// The max() protects against the clock getting set back
			$elapsed += max( $checkEndTime - $checkStartTime, 0.010 );
			// Do not let slow callbacks timeout without checking the condition one more time
			$lastCheck = ( $elapsed >= $this->timeout );
		} while ( true );

		$this->lastWaitTime = $elapsed;

		return $finalResult;
	}

	/**
	 * @return float Seconds
	 */
	public function getLastWaitTime() {
		return $this->lastWaitTime;
	}

	/**
	 * @param int $microseconds
	 * @codeCoverageIgnore
	 */
	protected function usleep( $microseconds ) {
		usleep( $microseconds );
	}

	/**
	 * @return float
	 * @codeCoverageIgnore
	 */
	protected function getWallTime() {
		return microtime( true );
	}

	/**
	 * @return float Returns 0.0 if not supported (Windows on PHP < 7)
	 * @codeCoverageIgnore
	 */
	protected function getCpuTime() {
		if ( $this->rusageMode === null ) {
			// assume worst case (all time is CPU)
			return microtime( true );
		}

		$ru = getrusage( $this->rusageMode );
		$time = $ru['ru_utime.tv_sec'] + $ru['ru_utime.tv_usec'] / 1e6;
		$time += $ru['ru_stime.tv_sec'] + $ru['ru_stime.tv_usec'] / 1e6;

		return $time;
	}

	/**
	 * Run one of the callbacks that does work ahead of time for another caller
	 *
	 * @return bool Whether a callback was executed
	 */
	private function popAndRunBusyCallback() {
		if ( $this->busyCallbacks ) {
			reset( $this->busyCallbacks );
			$key = key( $this->busyCallbacks );
			/** @var callable $workCallback */
			$workCallback =& $this->busyCallbacks[$key];
			try {
				$workCallback();
			} catch ( \Exception $e ) {
				$workCallback = static function () use ( $e ) {
					throw $e;
				};
			}
			// consume
			unset( $this->busyCallbacks[$key] );

			return true;
		}

		return false;
	}
}
